docs/signature-layout.md

```diff
@@ -0,0 +1,35 @@
+# Signature File Layout
+
+A common file layout for storing and serving signatures provides a consistent way to reference image signatures. Signatures on a filesystem or a web server shall use this common layout. Signatures stored in a REST API are not required to use this common layout.
+
+## Specification
+
+This specification relies on [RFC3986](https://tools.ietf.org/html/rfc3986), focusing on defining a [path component](https://tools.ietf.org/html/rfc3986#section-3.3) to compose a concise URI reference to a signature.
+
+**SCHEME[AUTHORITY]/PATH_PREFIX/IMAGE@MANIFEST_DIGEST/signature-INT**
+
+**Definitions**
+
+* **SCHEME**: URI scheme per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.1), e.g. **file://** or **https://**
+* **AUTHORITY**: An optional authority reference per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.2), e.g. **example.com**
+* **PATH_PREFIX**: An arbitrary base path to the image component
+* **IMAGE**: The name of the image per [v2 API](https://docs.docker.com/registry/spec/api/#/overview). This would typically take the form of registry/repository/image but is not required to have exactly three parts. There is no requirement to include a **:PORT** component but it should be included if part of the image reference.
```

Comment:
> >> … on-premises registry at reg.hrivnak.org. ¸ a local mirror … includes repositories like redhat/rhel7. A local client tries to pull reg.hrivnak.org/redhat/rhel7:latest …
>
> …The spec defined by this PR requires that the client insert the hostname of the original Red Hat registry into the URL. The client knows nothing about that registry, right?. Unless the intent is that all signatures on ``reg-sigstore.hrivnak.org`` would be namespaced to ``reg.hrivnak.org``?

Yes. As you say, the client cannot trivially determine that this is a mirror, it only knows the hostname/repo used to refer to the mirrored content.

> But if that's the case, why include a hostname at all?

Because it allows setting up a single sigstore server for multiple corporate registries.

Admittedly that is a fairly weak reason, but with the model of sigstores namespaced per-repo, not flat, within a single host name, removing the host name component would still not allow sharing signatures between differently-named repos. So, including the host name is a more symmetric / cleaner way to support the model in which the signatures are definitely namespaced, not shared, it prevents accidental unwanted sharing.

It can also be useful for a very locked-down site, where every allowed image needs to be individually approved with a company-private signature, to define a single sigstore for the whole internet; the image references would not change (so an upstream application which pulls `library/busybox:latest` does not need changing to point at a mirror), but the approving IT department would set up a single sigstore server “for the whole internet” serving these company-private signatures. Each host in the data center would be configured to require these private signatures, and to use this sigstore, but otherwise the applications can continue to use upstream references. [This is, more or less, an exact opposite of the “we mirror image layers but want to contact the original upstream for signatures; see below for more on that.]


>> there would need to be a single world-wide sigstore. 
>
>No, not at all!
>
>The idea is that any individual sigstore could host whatever collection of signatures are valuable to its audience of users. If you are an authority, you can host a sigstore with your signatures. If you want to mirror one or more authorities, you mirror the signature files from those authoritative sigstores. There's nothing centralized about it.

Hum, OK. So if I understand correctly you are considering a scenario where the user really wants to mirror the layers only because they are big and slow, and conceptually is perfectly fine with not having a mirror and fetching data directly.

If _that_ is the desire, a transparent proxy (caching both layers and signatures invisibly to the clients, as/if necessary) might be a cleaner way to achieve that: it means fewer modifications in the clients, OTOH a different kind of administrative effort/overhead.

In other cases, “an authority hosting signatures” is not a natural way to think about signatures. If there should be an authority saying “yes this image is OK / no I don’t know this one”, persistent signature objects are not needed: it can be an API provided by the authority over HTTPS.

One very important way to use signatures is to provide “non-repudiation”, i.e. a persistent, verifiable record that somebody has signed something. If an ISV publishes a backdoored image, and their client installs it because it has been signed by the ISV, it is very valuable for the client to later be able to attribute the image to the ISV even if the ISV tries to deny this (stops publishing the image and signatures). In that case it is essential for the client to have a _local_ copy of the signature, (along with a local copy of the signed image, or a local copy of enough other data to prove what the signature applies to).

And of course the client never knows that they will need to do such an investigation in advance. So, basically, the _standard operating procedure_ should be that any time an image is copied, the signatures are copied along, and perhaps logged somewhere. Making a mirror of a remote repo? Mirror the signatures. Pulling a container image onto a cluster host? Copy the signatures. Extracting the container image and preparing to run the container? Also record the signatures, at least in a log, or preferably in a form allowing to verify the extracted image against the signature.

Everywhere the image goes, the signatures go. A separate distribution mechanism with different lifetimes and under a different control is a risk to the ability to conduct forensic investigations.

>> I think the conceptual hurdle is to ignore how the sigstore technically works and to see that containers/image generally treats manifest+layers+signatures as single atomic unit. Then the philosophy and access control issues etc. follow naturally.
>
>The reason for this PR is to define aspects of how the sigstore technically works. I don't know what to do with your suggestion to ignore that.

(I’m honestly not sure how much this PR is still free to design things, or whether it is just to document the already shipped implementation.)

The thing is, containers/image can only work and be useful if the philosophy is consistent between various ways to store and access signatures; that `ImageDestination.PutSignatures` does something reasonably consistent between `ImageDestination` implementations. Consider that a design constraint if you like. (Of course it can be argued that the c/image API is wrong, and yes we can and do change it. But if `ImageDestination` becomes a mess of optional methods and every client will basically need to individually implement special signature storage semantics for every transport, c/image loses a lot of its value.)

>> (FWIW note that naive mirroring signatures with names cannot “just work” because the signatures claim the original (redhat/rhel7) identities, so pulling the reg.hrivnak.org/redhat/rhel7:latest will see an identity mismatch. That is quite unrelated to the sigstore mechanism, and we are prepared to support that, see hostname:5000/vendor/product in https://github.com/containers/image/blob/master/docs/policy.json.md#examples .)
>
>This is very interesting. Thanks.
>
>But per https://github.com/containers/image/pull/59/files#diff-c44796aa0013711d899f7345d01b8185R37, the signature would only include the name of the repository, and not include a registry hostname. I'm not sure where the identity mismatch would happen. Or is that PR (which is still not merged as of this writing) not accurate?

That does seem to be mistaken. It is “a string with the semantics and normalization conventions of `containers/image/docker/reference.Named`” (~upstream, not RH, `docker/docker/reference.Named`)”, and those strings do semantically specify host names, although the host name may be implicit.